### PR TITLE
Export chart to CSV [12025]

### PR DESCRIPTION
**Pull request by jparisu**
Merge after #92

### CODE REVIEW

File: qml/StatisticsChartBox.qml [hunk 1]
```diff
@@ -174,6 +175,13 @@ Rectangle {
                         statisticsChartBox.destroy()
                     }
                 }
+                Action {
+                    text: "Export all to CSV"
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:179]:
```suggestion
                    text: "Export to CSV"
```

File: qml/StatisticsChartBox.qml [hunk 2]
```diff
@@ -174,6 +175,13 @@ Rectangle {
...
+                    onTriggered: {
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:180]: The application freezes when saving a csv while a dynamic chart is running.

```suggestion
                    enabled: (isDynamic && !running) || (!isDynamic)
                    onTriggered: {
```

File: qml/StatisticsChartBox.qml [hunk 3]
```diff
@@ -201,6 +209,28 @@ Rectangle {
...
                         running ? statisticsChartViewLoader.item.dynamicPause() : statisticsChartViewLoader.item.dynamicContinue()
                     }
+
+                }
+            }
+        }
+
+        QtDialogs.FileDialog {
+            id: csvDialog
+            title: "Choose a file name and path to save csv"
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:219]:
```suggestion
            title: chartbox ? "Save " + chartTitle + " chart box data" :
                    "Save series " + customLegend.getLabel(seriesIndex) + " from " + chartTitle
```

File: qml/StatisticsChartBox.qml [hunk 4]
```diff
@@ -201,6 +209,28 @@ Rectangle {
...
+            folder: shortcuts.home
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:220]: The documents folder is more used by many users.
```suggestion
            folder: shortcuts.documents
```

File: qml/StatisticsChartBox.qml [hunk 5]
```diff
@@ -201,6 +209,28 @@ Rectangle {
...
+            selectMultiple: false
+            selectExisting: false
+            selectFolder: false
+            defaultSuffix: ".csv"
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:224]:
```suggestion
            defaultSuffix: ".csv"
            nameFilters: [ "CSV files (*.csv)", "All files (*)" ]
```

File: qml/CustomLegend.qml [hunk 6]
```diff
@@ -220,4 +225,16 @@ Rectangle {
     function removeSeries(seriesIndex) {
         seriesRemoved(seriesIndex);
     }
+
+    function getLabel(seriesIndex) {
+        return seriesModel.get(seriesIndex).seriesName
```
Comment (rsanchez15) [inline qml/CustomLegend.qml:230]: I'd append at the end of the series label the data units. For example for the latency it'll be something like `mean-part1-part2[us]` .

File: src/statistics/DataChartBox.cpp [hunk 7]
```diff
@@ -225,3 +226,165 @@ void DataChartBox::newXValue(
...
+        QString& file_name,
+        QVector<QString>& label_names,
+        std::string separator /* = ";"*/)
+{
+    const std::lock_guard<std::recursive_mutex> lock(mutex_);
+
+    assert(static_cast<size_t>(label_names.size()) == series_.size());
+
+    std::vector<QVector<QPointF>> datas(label_names.size());
+    // auto d = series_[0]->get_data();
```
Comment (rsanchez15) [inline src/statistics/DataChartBox.cpp:264]:
```suggestion
```

File: src/Engine.cpp [hunk 8]
```diff
@@ -642,6 +641,7 @@ QtCharts::QVXYModelMapper* Engine::on_add_statistics_data_series(
         end_time_default
             ? std::chrono::duration_cast<std::chrono::milliseconds>(
             time_to.time_since_epoch()).count()
+
```
Comment (rsanchez15) [inline src/Engine.cpp:644]:
```suggestion
```

File: src/main.cpp [hunk 9]
```diff
@@ -29,6 +29,9 @@ int main(
     QApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
     QApplication app(argc, argv);
     app.setWindowIcon(QIcon(":/resources/images/eprosima_logo.ico"));
+    app.setOrganizationName("eProsima");
+    app.setOrganizationDomain("eprosima.com");
+    app.setApplicationName("FastDDS Monitor");
```
Comment (rsanchez15) [inline src/main.cpp:34]:
```suggestion
    app.setApplicationName("Fast DDS Monitor");
```

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: qml/CustomLegend.qml (hunk 6 above)
Comment (rsanchez15) [reply qml/CustomLegend.qml:230]: Refactor csv in order to have the following format:

| | <datakind> |
| | <chartbox_title> |
| ms | <data_unit> |
| PosixTime | <series_label> |
| time | data |

File: qml/StatisticsChartBox.qml (hunk 3 above)
Comment (jparisu) [reply qml/StatisticsChartBox.qml:219]: :+1:

File: qml/StatisticsChartBox.qml (hunk 5 above)
Comment (jparisu) [reply qml/StatisticsChartBox.qml:224]: This does not force the file name, but just shows in folder other files of same type

File: qml/StatisticsChartBox.qml (hunk 3 above)
Comment (jparisu) [reply qml/StatisticsChartBox.qml:219]: Title in FileDialog is not visible, dont know why.

File: qml/StatisticsChartBox.qml (hunk 2 above)
Comment (jparisu) [reply qml/StatisticsChartBox.qml:180]: :+1:

File: qml/StatisticsChartBox.qml (hunk 3 above)
Comment (rsanchez15) [reply qml/StatisticsChartBox.qml:219]: This is the name of the new window opened.

File: qml/StatisticsChartBox.qml (hunk 5 above)
Comment (rsanchez15) [reply qml/StatisticsChartBox.qml:224]: I know but imagine that you want to replace another file. Selecting the option to see just the .csv files is easier to locate the file you are looking for. Also it's useful to see what .csv files you've already saved.

File: qml/StatisticsChartBox.qml [hunk 10]
```diff
@@ -216,12 +217,14 @@ Rectangle {
 
         QtDialogs.FileDialog {
             id: csvDialog
-            title: "Choose a file name and path to save csv"
-            folder: shortcuts.home
+            title: "TITLE" // chartbox ? "Save " + chartTitle + " chart box data" :
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:220]: ???

File: qml/StatisticsChartBox.qml (hunk 3 above)
Comment (rsanchez15) [reply qml/StatisticsChartBox.qml]: You missed this

File: include/fastdds_monitor/statistics/DataChartBox.h [hunk 11]
```diff
@@ -99,6 +99,11 @@ class DataChartBox : public QObject
     void newXValue(
             quint64 x);
 
+    const QVector<QPointF>& get_data(
+            quint64 series_index);
+
+    // std::vector<QVector<QPointF>> get_data();
```
Comment (rsanchez15) [inline include/fastdds_monitor/statistics/DataChartBox.h:105]: Don't forget to remove

File: src/Controller.cpp [hunk 12]
```diff
@@ -210,3 +210,25 @@ void Controller::set_alias(
...
+    return utils::to_QString(engine_->get_data_kind_units(data_kind));
+}
+
+void Controller::save_csv(
+        QString file_name,
+        QVector<quint64> chartbox_ids,
+        QVector<quint64> series_indexes,
+        QVector<QString> data_kinds,
+        QVector<QString> chartbox_names,
+        QVector<QString> label_names)
```
Comment (rsanchez15) [inline src/Controller.cpp:225]: Remember to change to QList<...>

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: src/Controller.cpp (hunk 12 above)
Comment (jparisu) [reply src/Controller.cpp]: :+1:

**Comment by jparisu** [comment]
Ready to merge with CI in green and one approval.
Please, check vectors are not copied along the export csv execution.

File: qml/ChartsLayout.qml [hunk 13]
```diff
@@ -210,5 +216,52 @@ Rectangle {
     function calculateGridViewWidth(){
         mouseArea.width = (chartBoxWidth - 10) * actualBoxesPerRow
     }
+
+    function save_all_csv(){
+        // TODO
+    }
```
Comment (rsanchez15) [inline qml/ChartsLayout.qml:222]: Duplicated function. Remove.

File: qml/MonitorMenuBar.qml [hunk 14]
```diff
@@ -29,6 +30,11 @@ MenuBar {
             onTriggered: dialogInitMonitor.open()
         }
         MenuSeparator { }
+        Action {
+            text: qsTr("Export all Charts to &CSV")
```
Comment (rsanchez15) [inline qml/MonitorMenuBar.qml:34]:
```suggestion
            text: qsTr("Save Charts in &CSV")
```

File: qml/StatisticsChartBox.qml [hunk 15]
```diff
@@ -18,6 +18,7 @@ import QtQuick.Window 2.15
 import QtQuick.Layouts 1.3
 import QtGraphicalEffects 1.15
 import Theme 1.0
+import QtQuick.Dialogs 1.3 as QtDialogs
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:21]:
```suggestion
```

File: qml/StatisticsChartBox.qml [hunk 16]
```diff
@@ -201,6 +218,7 @@ Rectangle {
                     onTriggered: {
                         running ? statisticsChartViewLoader.item.dynamicPause() : statisticsChartViewLoader.item.dynamicContinue()
                     }
+
```
Comment (rsanchez15) [inline qml/StatisticsChartBox.qml:221]:
```suggestion
```

File: qml/StatisticsChartView.qml [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import QtQuick 2.0
+import QtCharts 2.1
+import QtQuick.Controls 2.15
+
+ChartView {
+    id: sharedChartView
+    animationOptions: ChartView.NoAnimation // Some animations misalign the Timer period (NPI)
```
Comment (rsanchez15) [inline qml/StatisticsChartView.qml:21]: :disappointed:

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: qml/ChartsLayout.qml (hunk 13 above)
Comment (jparisu) [reply qml/ChartsLayout.qml:222]: :+1:

File: CMakeLists.txt [hunk 18]
```diff
@@ -210,6 +210,7 @@ set(PROJECT_HEADERS
     ${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}/backend/SyncBackendConnection.h
     ${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}/Controller.h
     ${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}/Engine.h
+    ${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}/io/csv.h
```
[the remark this answers is not in the document]
Comment (jparisu) [reply CMakeLists.txt:213]: :+1:

File: CMakeLists.txt [hunk 19]
```diff
@@ -243,6 +244,7 @@ set(PROJECT_SOURCES_NO_MAIN
     ${PROJECT_SOURCE_DIR}/src/backend/SyncBackendConnection.cpp
     ${PROJECT_SOURCE_DIR}/src/Controller.cpp
     ${PROJECT_SOURCE_DIR}/src/Engine.cpp
+    ${PROJECT_SOURCE_DIR}/src/io/csv.cpp
```
[the remark this answers is not in the document]
Comment (jparisu) [reply CMakeLists.txt:247]: :+1:

File: fastdds_monitor.pro [hunk 20]
```diff
@@ -13,6 +13,7 @@ SOURCES += \
         src/backend/SyncBackendConnection.cpp \
         src/Controller.cpp \
         src/Engine.cpp \
+        src/io/csv.cpp \
```
[the remark this answers is not in the document]
Comment (jparisu) [reply fastdds_monitor.pro:16]: :+1:

File: fastdds_monitor.pro [hunk 21]
```diff
@@ -56,6 +57,7 @@ HEADERS += \
     include/fastdds_monitor/backend/SyncBackendConnection.h \
     include/fastdds_monitor/Controller.h \
     include/fastdds_monitor/Engine.h \
+    include/fastdds_monitor/io/csv.h \
```
[the remark this answers is not in the document]
Comment (jparisu) [reply fastdds_monitor.pro:60]: :+1:

**Review by rsanchez15 — APPROVED** [review]
LGTM :rocket:

**Pull request merged**